tokio-console/src/state/mod.rs

```diff
@@ -218,16 +229,41 @@ impl State {
 
     pub(crate) fn update_task_details(&mut self, update: proto::tasks::TaskDetails) {
         if let Some(id) = update.task_id {
-            let details = Details {
-                span_id: id.id,
-                poll_times_histogram: update.poll_times_histogram.and_then(|data| {
-                    hdrhistogram::serialization::Deserializer::new()
-                        .deserialize(&mut Cursor::new(&data))
-                        .ok()
-                }),
+            let span_id = id.id;
+            let task_id = self.tasks_state.ids.id_for(span_id);
+            let task_ref = self.tasks_state.task(task_id).unwrap();
+            let task_ref = task_ref.upgrade().unwrap();
+            let meta_id = task_ref.borrow().meta_id();
+            let poll_times_histogram = update.poll_times_histogram.and_then(|data| {
+                hdrhistogram::serialization::Deserializer::new()
+                    .deserialize(&mut Cursor::new(&data))
+                    .ok()
+            });
+            let details = self.current_task_details.take();
+            let mut fresh = Details {
+                span_id,
+                poll_times_histogram,
+                causality: details.and_then(|details| details.causality),
             };
 
-            *self.current_task_details.borrow_mut() = Some(details);
+            if let None = fresh.causality {
+                fresh.causality = Some(tracing_causality::Trace::with_root(
+                    tracing_causality::Span {
+                        id: tracing::Id::from_u64(span_id),
+                        metadata: meta_id,
+                    },
+                ));
+            }
+
+            for causal_update in update.causality {
+                for update in trace::updates_from_pb(causal_update.update.unwrap()) {
+                    if let Some(trace) = fresh.causality {
+                        fresh.causality = trace.apply(update);
+                    }
```

Comment:
nit, take it or leave it: this could be:
```suggestion
                    fresh.causality = fresh.causality.and_then(|trace| trace.apply(update));
```